### ISSUE TITLE
Implement CPU forward + KV cache (P0)

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#462]
# Post-merge read & next steps (to MVP + “drop-in CPU”)

## Where we are (after #461)

* **Strict quantization guards** are merged: FP32 staging is blocked under `BITNET_STRICT_MODE=1`; receipts verify kernel honesty.
* **What’s still missing for MVP answers:** a real **CPU forward path** (with KV cache) wired to the CLI so `bitnet run …` emits **actual tokens**. Throughput gate was marked N/A—confirming inference surface still needs to land.

---

### 1) Implement CPU forward + KV cache (P0)

**Why:** `forward_*` still produces placeholder logits; nothing else matters until this is real.

**Spec (single-step decode, B=1):**

* `embed(token)` → `x ∈ [1, d_model]`
* For each block: `x = x + O(attn(Q/K/V, KV_cache))`; `x = x + FFN(LN(x))`
* `logits = LM_HEAD(LN(x)) ∈ [1, vocab]`
* KV cache per layer: `K,V ∈ [H, T, Dh]`; append per token; scale = `1/√Dh`.

**DoD**

* Unit: BOS → **non-zero logits** (finite; ≠0).
* Integration: greedy **16-token** decode runs (F16 model).
* Strict on: no FP32 fallbacks taken on quantized layers.

---

### 2) Wire CLI `run` to real logits (P0)

**Why:** Priming + sampling still routes through the old stub.

**DoD**

* Prime by feeding prompt tokens → loop {logits → sample (greedy/top-k/p) → next token → forward}.
* `Q: What is 2+2? A:` returns `4` (deterministic) in first tokens at `temperature=0.0`.

---

### 3) CPU receipt honesty + safer kernel matchers (P1)

**Why:** GPU receipts already enforce GPU kernels; add **CPU symmetry**. Also eliminate `contains(…)` false positives (e.g., “dequantize_*” matching “quantize_”).

**Change**

* Verifier requires **≥1** CPU quantized kernel prefix in receipts when `backend:"cpu"` (`i2s_`/`tl1_`/`tl2_`).
* Switch all kernel classifiers to **`starts_with`**; explicitly exclude `dequant*`, `fp32_*`, `fallback_*`; keep `== "matmul_f32"` as fallback.

**DoD**

* Positive CPU receipt passes only if a quantized kernel prefix is present.
* Negative CPU receipt (fallback-only) fails with actionable message.
* Existing GPU honesty tests continue to pass.

---

### 4) TL1/TL2 LUT indexing helper; re-enable TL tests (P1)

**Why:** Fix the off-by-block class that forced `#[ignore]`.

**Change**

* Introduce `lut_index(block_idx, elem_in_block, block_bytes, elems_per_block)` with bounds.
* Use at all TL1/TL2 lookup sites; re-enable TL tests.

**DoD**

* TL tests flip to **green** on CPU.

---

### 5) AC8 GPU baseline — real, opt-in, skip-clean (P2, optional for MVP)

**Why:** Keep CI fast, but make the acceptance test meaningful on CUDA hosts.

**Change**

* Guard the test on `feature="gpu"` and `BITNET_ENABLE_GPU_TESTS=1`; **skip** cleanly if no CUDA at runtime.
* Run a 128-token GPU bench → build an in-memory receipt → check **GPU kernel IDs** + per-fingerprint throughput envelope (`.ci/fingerprints.yml`).

**DoD**

* Non-CUDA/without env → **skipped** (neutral).
* CUDA + env → runs and verifies receipt honesty + envelope.

---

### 6) Pin a baseline receipt + README quickstart (P1)

**Why:** Make “evidence over prose” visible.

**Change**

* Run `xtask benchmark` (strict+deterministic) on tiny GGUF; copy `ci/inference.json` → `docs/baselines/YYYYMMDD-cpu.json`.
* README: **one copy-paste** (“question → answer”) and a **receipt-first** box (“run → emit → verify”).

**DoD**

* Quickstart produces an answer on F16; verifier passes; pinned receipt shows `compute_path:"real"` and CPU quantized kernels.

---

### 7) Flip “Model Gates (CPU)” to **required** on `main` (post-merge admin)

**DoD**

* A PR with bad receipt (e.g., `compute_path:"mock"`) is blocked by the required job.

---

## Ready-to-open issues / PRs (copy-paste)

1. **feat(inference): implement CPU forward_step + KV cache**
   *Why:* unblock real tokens; strict guarantees quantized hot-path.
   *DoD:* non-zero logits on BOS; 16-token greedy decode; strict on; CPU tests green.

2. **feat(cli): wire run to real logits; prime + greedy/top-k/p**
   *DoD:* demo prompt returns coherent tokens; deterministic at temp 0.0.

3. **feat(receipts): require CPU quantized kernel prefixes; safe prefix matchers**
   *DoD:* positive/negative CPU tests; fix `contains → starts_with`; exclude `dequant*/fp32_/fallback_`.

4. **fix(tl): bounds-checked TL LUT index; re-enable TL tests**
   *DoD:* TL tests green; no bounds assertions hit in debug.

5. **test(ac8): GPU baseline receipt — GPU-only, skip-clean, allowlist** *(optional for MVP)*
   *DoD:* opt-in test runs on CUDA; skipped elsewhere; receipt honesty + envelope verified.

6. **docs(baselines): pin CPU receipt; README quickstart + receipts how-to**
   *DoD:* pinned JSON; copy-paste quickstart works; link to verifier.

7. **ops(ci): require Model Gates (CPU) on main**
   *DoD:* branch protection enabled; smoke PR proves gating.

---

## Verifier commands (quick runbook)

```bash
# Strict + deterministic
export BITNET_STRICT_MODE=1 BITNET_DETERMINISTIC=1 RAYON_NUM_THREADS=1

# 1) Tests (CPU)
cargo test --workspace --no-default-features --features cpu

# 2) Bench + receipt verify (CPU)
cargo run -p xtask -- benchmark --model tests/models/tiny.gguf --tokens 128 --deterministic
cargo run -p xtask -- verify-receipt ci/inference.json

# 3) GPU AC8 (optional, on CUDA host)
export BITNET_ENABLE_GPU_TESTS=1
cargo test -p bitnet-inference --features gpu -- ac8_gpu_performance_baselines -q
```

---

## Risks & knobs

* **Throughput SLO (≤10s for 128)**: On some CPUs, 128 tokens might exceed 10s. If so, validate SLO with **100 tokens** (document the bound) while keeping receipts for 128.
* **Tokenizer discovery**: always support explicit `--tokenizer` override; fail fast on ambiguity.
* **Strict failures**: clear error strings (layer name + “FP32 staging fallback blocked by strict”).

---

## Effort to finish (solo)

* **CPU forward + CLI wiring:** ~**1–2 pd**
* **Receipts & TL fixes:** ~**0.5–1 pd**
* **Docs + baseline + gate flip:** ~**0.5 pd**
* **(Optional) AC8 GPU test:** ~**0.5–1 pd**

---

If you want, I can draft the three core diffs as a single PR (`feat(inference): cpu forward + cli wiring; feat(receipts): cpu honesty + prefix matchers; fix(tl): lut helper + tests`) with unit/integration tests and a README quickstart update ready to merge.

# BitNet.rs — full implementation plan to cross MVP (connector-verified)

Below is a concrete, codebase-aligned plan to go from “strict + receipts merged” to **answers on CPU** and a credible **CPU drop-in**. Each task names the file(s) to touch, shows the exact seam in your repo, and states a tight DoD (definition of done). Citations link to the relevant code anchors.

---

## 0) Why this plan (what’s missing)

* The **CPU forward path** still returns placeholder logits (“zeros [1, 32000]”), so the CLI cannot produce real tokens. We must replace that stub with a real one-step decode + KV cache. 
* The **quantized linear** layer already selects native quantized kernels or falls back to dequant (we’ll keep strict guards hard-blocking fallbacks).  
* The **CLI** can load model & tokenizer and create the engine; we need to wire priming → forward → sampling to the **real** logits. 

---

## 1) Implement **CPU forward** (+ KV cache) — P0

**File(s):** `crates/bitnet-inference/src/cpu.rs` (engine)
**Current:** `CpuInferenceEngine::forward_parallel(...)` returns placeholder zeros (doc/spec shows the stub code and file path).  

**Change:** implement a one-step decode:

* `embed(token)` → `x: [1, d_model]`
* For each layer:

  * `xn = LN(x)`
  * **Q/K/V** via **QuantizedLinear** (I2S/TL1/TL2) — *strict* must refuse FP32 staging; this is already the contract in the PR. 
  * append **K,V** to `KVCache[layer]: [H, T, Dh]`
  * **attention**: single-step matmul (per-head OK at MVP), scale `1/√Dh`, causal by construction
  * **O** via **QuantizedLinear**, add residual
  * **FFN** (pre-norm), add residual
* `logits = LM_HEAD(LN(x))` → `[1, vocab]`

> The quantized linear forward paths are present and structured to select native kernels or fall back; we keep strict guards to block dequant fallbacks in this hot path.  

**DoD:**

* Unit: BOS → logits **finite and non-zero**.
* Integration: **16-token greedy** decode runs on clean F16 GGUF without panic.

---

## 2) Wire **CLI** to real logits — P0

**File:** `crates/bitnet-cli/src/commands/inference.rs`
**Current:** can load model/tokenizer and build `InferenceEngine`. 

**Change:**

* **Priming:** feed all prompt tokens one-by-one into `engine.forward_parallel(..)` (or `forward_step(..)` if you add it), updating KV cache.
* **Decode loop:** logits → `sample_greedy` (temp=0) or `top-k/p` → next token → forward → print.

**DoD:**

* `bitnet run --model … --prompt "Q: What is 2+2? A:" --max-new-tokens 16 --temperature 0.0` yields `4` in first tokens.

---

## 3) Receipt honesty (CPU symmetry) & safe matchers — P1

**Files:** `xtask/src/verify_receipt.rs` (or existing verifier) and tests in `xtask`
**Current:** GPU verification requires GPU-kernel prefixes; CPU rule is looser. A matcher using `contains()` risks “de**quantize****” matching “**quantize***”. Your receipt tests already rely on GPU prefix validity (e.g., `cuda` backend vs CPU kernels). 

**Change:**

* **CPU rule:** require at least one **CPU quantized kernel prefix** (`i2s_`, `tl1_`, `tl2_`) when `backend:"cpu"`.
* **Prefix only:** replace any `contains()` checks with **`starts_with()`** and **exclude** `dequant*`, `fp32_*`, `fallback_*`; keep `== "matmul_f32"` as explicit fallback classification.

**DoD:**

* A CPU receipt **without** `i2s_/tl*_` fails with a clear error.
* Existing GPU negative test (CUDA backend with CPU kernels) stays failing. 

---

## 4) TL1/TL2 LUT index helper; flip TL tests to green — P1

**Files:**

* **New:** `crates/bitnet-kernels/src/tl_lut.rs` (`lut_index(block_idx, elem_in_block, block_bytes, elems_per_block)` with bounds)
* **Change:** call sites in `crates/bitnet-inference/src/layers/quantized_linear.rs` TL1/TL2 generic/vector paths to use the helper.

**Why:** TL fallbacks currently dequant + matmul; vectorized paths pick a provider if available. Off-by-block math caused prior ignores; bound the math and re-enable TL tests.  

**DoD:** TL tests in `crates/bitnet-inference/tests` move from `#[ignore]` to **pass**.

---

## 5) AC8 GPU baseline — **receipt-driven**, GPU-only, skip-clean — P2 (optional for MVP)

**File:** `crates/bitnet-inference/tests/issue_260_mock_elimination_inference_tests.rs`
**Current:** AC8 module exists with a CUDA device check; make it meaningful but opt-in. 

**Change:**

* Gate test with `#[cfg(feature="gpu")]` **and** env `BITNET_ENABLE_GPU_TESTS=1`.
* Skip cleanly if `Device::new_cuda(0)` fails.
* Run a short 128-token GPU bench, build an **in-memory receipt**, assert:

  * **GPU kernel honesty** (≥1 GPU kernel ID prefix such as `gemm_`, `wmma_`, `cuda_`, `i2s_gpu_`, `tl*_gpu_`).
  * **Throughput envelope**: default 50–100 tok/s; widen per-fingerprint via `.ci/fingerprints.yml`.

**DoD:**

* Non-CUDA/without env → **skipped** (neutral).
* CUDA + env → **passes** with receipt checks.

---

## 6) Pin a **baseline receipt** + README quickstart — P1

**Files:**

* `docs/baselines/` (new JSON copied from `ci/inference.json`)
* `README.md` (quickstart + receipts section)

**Change:**

* Strict+deterministic bench on tiny GGUF → `ci/inference.json` → copy to `docs/baselines/YYYYMMDD-cpu.json`.
* Add a 10-line “**question → answer**” block (F16), and a “**run → emit → verify**” block mirroring your CLI/xtask docs. (The CLI inference command and “deterministic” envs are already documented and used elsewhere.  )

**DoD:**

* Copy-paste quickstart **produces an answer**.
* `xtask verify-receipt` **passes**, and the pinned file shows `compute_path:"real"` + CPU quantized kernels.

---

## 7) Post-merge: turn on the **required** CPU gate

**Action:** Repo settings → Branch protection for `main` → require the **Model Gates (CPU)** job.

**DoD:** A PR with a bad receipt (e.g., `compute_path:"mock"`) is **blocked**.

---

# Implementation map — files & seams

* **Engine:** implement real forward & KV cache in `crates/bitnet-inference/src/cpu.rs`, replacing the zero-logits stub referenced in your issue doc.  
* **Quantized layers:** reuse `QuantizedLinear::{forward_i2s, forward_tl1, forward_tl2}`; strict guard remains the policy seam.  
* **TL LUT helper:** add helper in `bitnet-kernels/src/tl_lut.rs`; replace ad-hoc math in `quantized_linear.rs` TL paths. 
* **CLI run:** wire priming + sampling in `crates/bitnet-cli/src/commands/inference.rs`. 
* **Strict:** continue using `StrictModeConfig::{require_quantization, …}`; the enforcer API is already stable. 
* **Receipts:** extend the CPU rule and tighten matchers in `xtask` (your existing tests already validate GPU kernel honesty). 

---

# Test plan (add/flip)

1. **Unit (CPU forward):**

   * `f16_cpu_forward_nonzero_logits` (BOS produces non-zero; finite).
   * `greedy_decode_16_tokens` (no panic; tokens stream).

2. **TL tests:**

   * Re-enable TL1/TL2 indexing tests after the LUT helper lands. (Vectorized path may still call generic; strict only blocks when required.)

3. **Receipts (xtask):**

   * Positive/negative CPU receipts (with/without `i2s_/tl*_`).
   * Ensure GPU negative case (CUDA backend + CPU kernels) keeps failing. 

4. **AC8 (optional):**

   * Add **opt-in** GPU baseline receipt test (skip-clean by default). 

---

# Throughput gate (practical SLO)

* Use tiny GGUF to stay ≤ 10s for 128 tokens; if your CPU cannot, validate on 100 tokens and document the bound in the gate summary (SLO: ≤10s for **standard** tiny model).

---

# Burndown (solo)

* **CPU forward + KV + CLI wiring:** 1–2 pd
* **Receipts CPU rule + matcher fix:** 0.5 pd
* **TL LUT helper + flip tests:** 0.5–1 pd
* **Docs + baseline + required gate:** 0.5 pd
* **(Optional) AC8 GPU test:** 0.5–1 pd

---

## Receipts you can paste as you go

* **Engine seam (stub):** “`forward_parallel` creates placeholder zeros; implement real decode.” 
* **Quantized paths:** “native quantized or explicit fallback paths in `QuantizedLinear`.”  
* **CLI hook:** “engine build/wiring in `inference.rs`.” 
* **Strict enforcer API:** “validate kernel availability / no mock use.” 
* **GPU receipt tests exist:** extend to CPU symmetry. 

---

If you want, I can split this into two PRs:

1. **feat(inference): CPU forward + CLI wiring + TL LUT helper (+ tests)**
2. **feat(receipts): CPU honesty + prefix matchers; test(ac8): GPU skip-clean; docs(baseline + quickstart)**

Either route gets you to **MVP answers on CPU** with receipts and strict guarantees, and puts you a single thin wrapper away from “drop-in for bitnet.cpp (CPU)”.
